### PR TITLE
IGNITE-14465: Add the ability to set and get cluster state

### PR DESCRIPTION
**Pull request by isapego**

### CODE REVIEW

File: pyignite/queries/response.py [hunk 1]
```diff
@@ -44,7 +45,7 @@ def __build_header(self):
                 ('query_id', ctypes.c_longlong),
             ]
 
-            if self.protocol_version and self.protocol_version >= (1, 4, 0):
+            if self.protocol_context and self.protocol_context.is_status_flags_supported():
```
Comment (ivandasch) [inline pyignite/queries/response.py:48]: I suppose, that check for None is ambigous here.... Let's remove it (if we are here, protocol context have been already set)

File: pyignite/connection/protocol_context.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple
+
+from pyignite.connection.bitmask_feature import BitmaskFeature
+
+
+class ProtocolContext:
```
Comment (ivandasch) [inline pyignite/connection/protocol_context.py:21]: Let's make it hashable (`__hash__` and `__eq__`)  If we do it, we can cache c_types structs in dict by specific context.

File: pyignite/connection/connection.py [hunk 3]
```diff
@@ -180,22 +166,22 @@ def connect(self) -> Union[dict, OrderedDict]:
...
             detecting_protocol = True
-            self.client.protocol_version = max(PROTOCOLS)
+            self.client.protocol_context = ProtocolContext(max(PROTOCOLS), BitmaskFeature.all_supported())
 
         try:
             result = self._connect_version()
         except HandshakeError as e:
             if e.expected_version in PROTOCOLS:
-                self.client.protocol_version = e.expected_version
+                self.client.protocol_context.version = e.expected_version
```
Comment (ivandasch) [inline pyignite/connection/connection.py:177]: We should discard `protocol_context.features` also.

File: pyignite/connection/connection.py [hunk 4]
```diff
@@ -227,7 +213,7 @@ def _connect_version(self) -> Union[dict, OrderedDict]:
             self.send(stream.getbuffer(), reconnect=False)
 
         with BinaryStream(self.client, self.recv(reconnect=False)) as stream:
-            hs_response = HandshakeResponse.parse(stream, self.protocol_version)
+            hs_response = HandshakeResponse.parse(stream, self.protocol_context)
```
Comment (ivandasch) [inline pyignite/connection/connection.py:216]: We should update from response `protocol_context.features`

File: pyignite/connection/aio_connection.py [hunk 5]
```diff
@@ -158,7 +160,7 @@ async def _connect_version(self) -> Union[dict, OrderedDict]:
             await self._send(stream.getbuffer(), reconnect=False)
 
         with AioBinaryStream(self.client, await self._recv(reconnect=False)) as stream:
-            hs_response = await HandshakeResponse.parse_async(stream, self.protocol_version)
+            hs_response = await HandshakeResponse.parse_async(stream, self.protocol_context)
```
Comment (ivandasch) [inline pyignite/connection/aio_connection.py:163]: We should update from response `protocol_context.features`

Comment (isapego) [reply pyignite/connection/aio_connection.py:163]: Good catch. Fixed.

File: pyignite/connection/connection.py (hunk 3 above)
Comment (isapego) [reply pyignite/connection/connection.py]: Fixed.

File: pyignite/connection/protocol_context.py (hunk 2 above)
Comment (isapego) [reply pyignite/connection/protocol_context.py:21]: Done

File: pyignite/queries/response.py (hunk 1 above)
Comment (isapego) [reply pyignite/queries/response.py:48]: Done

File: pyignite/connection/connection.py (hunk 4 above)
Comment (isapego) [reply pyignite/connection/connection.py:218]: I update it a bit later

File: pyignite/connection/protocol_context.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    def _ensure_consistency(self):
+        if not self.is_feature_flags_supported():
+            self._features = None
+
+    @property
+    def version(self):
+        return getattr(self, '_version', None)
+
+    def set_version(self, version: Tuple[int, int, int]):
```
Comment (ivandasch) [inline pyignite/connection/protocol_context.py:49]: Why not use `@version.setter` ?

File: tests/custom/conftest.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+import pytest
+
+from pyignite import Client, AioClient
+from tests.util import start_ignite
+
+
+@pytest.fixture(scope='module')
+def start_ignite_server():
+    def start(idx=1, debug=False, use_ssl=False, use_auth=False, use_persistence=False):
```
Comment (ivandasch) [inline tests/custom/conftest.py:24]: But who will kill ignite?

File: tests/custom/conftest.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            use_ssl=use_ssl,
+            use_auth=use_auth,
+            use_persistence=use_persistence,
+        )
+
+    return start
+
+
+@pytest.fixture(scope='module')
+def start_client():
```
Comment (ivandasch) [inline tests/custom/conftest.py:37]: Who will close client? Please, see another _conftest.py

File: pyignite/connection/protocol_context.py (hunk 6 above)
Comment (isapego) [reply pyignite/connection/protocol_context.py:49]: I'm not sure that it's OK when setter has some side-logic. Is it?

Comment (ivandasch) [reply pyignite/connection/protocol_context.py:49]: I suppose that it is main purpose of getters/setters :)

Comment (isapego) [reply pyignite/connection/protocol_context.py:49]: Changed it.

**Review by ivandasch — APPROVED** [review]

**Pull request closed without merging**